### PR TITLE
feat: Add device onboarding type 'preconnect' 

### PR DESCRIPTION
**Pull request by sherretten**
## Problem
We added a way to connect a single device to nRF Cloud via the  [Onboard Device API](https://api.nrfcloud.com/v1/#tag/IP-Devices/operation/OnboardDevice), and the simulator only could connect via the [Associate Device API](https://api.nrfcloud.com/v1/#tag/IP-Devices/operation/AssociateDevice) endpoint.

## Solution
1. Change the `-a` flag to allow for two options `jitp` or `preconnect`
2. With preconnect it will create a CA cert, then a CSR required for connection
3. Then it will call the Onboard device endpoint mentioned above.

## Testing
1. Checkout this branch
2. `yarn build`
3. Simulate a device with `-a preconnect` or `-a jitp` for the old way:
``` sh
node dist/cli.js -k <api-key> -d <deviceId> -h https://api.dev.nrfcloud.com -a preconnect -t atv2
```
4. Everything should look green :)

### Jira
https://nordicsemi.atlassian.net/browse/IRIS-8917

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -24,15 +24,15 @@ You can name the device whatever you want with the `-d` option. If not present,
...
 ```
 
 ### Simulate sensor outputs
 Include any combination of the options listed after `-s` below:
 ```
-npx @nrfcloud/device-simulator-v2 -k <api key> -a -d <device ID you already associated> -s gps,acc,device,temp
+npx @nrfcloud/device-simulator-v2 -k <api key> -a preconnect -d <device ID you already onboarded> -s gps,acc,device,temp
 ```
-*Note! Usually the `-a` is not necessary, since the device is already associated to your account. However, due to bug IRIS-3450, this does not always work. Adding the `-a` is harmless and is a workaround for now.*
+*Note! Usually the `-a` is not necessary, since the device is already onboarded to your account. However, due to bug IRIS-3450, this does not always work. Adding the `-a` is harmless and is a workaround for now.*
```
Comment (gfleming-nordic) [inline README.md]: is this still true? doesn't this change the command to now become necessary  (telling the simulator to use the "jitp" or "preconnect" modality). It also seems IRIS-3450 is not reproducible anymore?

File: src/cli.ts [hunk 2]
```diff
@@ -73,6 +73,21 @@ const handleJobExecution = (input: string, _: unknown) => {
...
+  }
+
+  if (input === 'jitp') {
+    return input;
+  }
+
+  new Log(false).error(
+    'Input for associate must be blank, "jitp" or "preconnect"',
+  );
+  process.exit();
```
Comment (gfleming-nordic) [inline src/cli.ts]: I'm picking nits but the returns are inconsistent here, line 78 returns a hardcoded string, while L82 returns the actual input, I would like to see that be consistent. Also, adding a `process.exit(1);` is just good form

could also be managed something like this way. 

```suggested
  const validInputs = ['preconnect', 'jitp'];

  if (validInputs.includes(input)) {
    return input;
  }

  new Log(false).error(
    'Input for onboarding type must be "jitp" or "preconnect"',
  );
  process.exit(1);
```

File: src/cli.ts [hunk 3]
```diff
@@ -73,6 +73,21 @@ const handleJobExecution = (input: string, _: unknown) => {
...
+  if (input === 'preconnect') {
+    return 'preconnect';
+  }
+
+  if (input === 'jitp') {
+    return input;
+  }
+
+  new Log(false).error(
+    'Input for associate must be blank, "jitp" or "preconnect"',
```
Comment (gfleming-nordic) [inline src/cli.ts]: I think 
~~Input for associate must be blank,~~
should become
`Input for onboarding type must be`

**Review by gfleming-nordic — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (lincolnthomas) [reply README.md]: In the series of sections above, the device is already onboarded (associated) in the "Onboard device to your account" section. But maybe it's not obvious that this is a series of commands. Since IRIS-3450 is not an issue anymore (we think!), you can just remove the Note. But let's ensure people reading know it has to be associated/onboarded already, see above.

File: README.md [hunk 4]
```diff
@@ -24,15 +24,15 @@ You can name the device whatever you want with the `-d` option. If not present,
 ### Associate device to your account
 This will create a new device and associate it to the account for the API key.
 ```
-npx @nrfcloud/device-simulator-v2 -k <api key> -a -t atv2
+npx @nrfcloud/device-simulator-v2 -k <api key> -a -t atv2 -a preconnect
 ```
 
 ### Simulate sensor outputs
 Include any combination of the options listed after `-s` below:
```
Comment (lincolnthomas) [inline README.md]:
```suggestion
After creating and associating your device, you can include any combination of the options listed after `-s` below, to simulate the device sending messages containing device sensor data:
```

File: README.md [hunk 5]
```diff
@@ -24,15 +24,15 @@ You can name the device whatever you want with the `-d` option. If not present,
 ### Associate device to your account
```
Comment (lincolnthomas) [inline README.md]:
```suggestion
### Onboard a device to your account
```

File: README.md [hunk 6]
```diff
@@ -24,15 +24,15 @@ You can name the device whatever you want with the `-d` option. If not present,
...
 This will create a new device and associate it to the account for the API key.
```
Comment (lincolnthomas) [inline README.md]:
```suggestion
This example will onboard a device to nRF Cloud by creating a new simulated device and associating it to the user's team. The user's API key authenticates the REST request. A default device shadow will be created in the style of the Asset Tracker v2 (atv2) sample firmware. For the pre-connect mode of onboarding, the device simulator will create and provide self-signed certificates in the onboarding process.
```

File: src/index.ts [hunk 7]
```diff
@@ -219,27 +293,37 @@ export const run = async (config: SimulatorConfig): Promise<void> => {
 
   log.success('starting simulator...');
 
-  if (associate) {
+  if (onboard) {
     config.onConnect = async (deviceId) => {
       log.info(
-        `ATTEMPTING TO ASSOCIATE ${config.deviceId} WITH API KEY ${config.apiKey} VIA ${config.apiHost}`,
+        `ATTEMPTING TO ONBOARD ${config.deviceId} via ${onboard} WITH API KEY ${config.apiKey} VIA ${config.apiHost}`,
```
Comment (lincolnthomas) [inline src/index.ts]:
```suggestion
        `ATTEMPTING TO ONBOARD ${config.deviceId} USING ${onboard} CERTS WITH API KEY ${config.apiKey} ON HOST ${config.apiHost}`,
```

File: README.md (hunk 5 above)
Comment (lincolnthomas) [reply README.md]: Also change some more "associate" --> "onboard" references earlier in this file.

File: README.md [hunk 8]
```diff
@@ -24,15 +24,15 @@ You can name the device whatever you want with the `-d` option. If not present,
...
 ```
-npx @nrfcloud/device-simulator-v2 -k <api key> -a -t atv2
+npx @nrfcloud/device-simulator-v2 -k <api key> -a -t atv2 -a preconnect
 ```
 
 ### Simulate sensor outputs
 Include any combination of the options listed after `-s` below:
 ```
-npx @nrfcloud/device-simulator-v2 -k <api key> -a -d <device ID you already associated> -s gps,acc,device,temp
+npx @nrfcloud/device-simulator-v2 -k <api key> -a preconnect -d <device ID you already onboarded> -s gps,acc,device,temp
```
Comment (lincolnthomas) [inline README.md]:
```suggestion
npx @nrfcloud/device-simulator-v2 -k <api key> -d <device ID you already onboarded> -s gps,acc,device,temp
```

File: src/index.ts [hunk 9]
```diff
@@ -219,27 +293,37 @@ export const run = async (config: SimulatorConfig): Promise<void> => {
...
       );
 
       // wait to ensure the device is available in AWS IoT so it can be associated
       await new Promise((resolve) => setTimeout(resolve, 2000));
```
Comment (lincolnthomas) [inline src/index.ts]: Should this waiting period change for preconnect @gfleming-nordic ? Is it even necessary?

**Review by lincolnthomas — CHANGES_REQUESTED** [review]

File: src/cli.ts (hunk 2 above)
Comment (sherretten) [reply src/cli.ts]: Good catch

File: src/index.ts (hunk 9 above)
Comment (sherretten) [reply src/index.ts]: Works right away without. Good catch!

Comment (gfleming-nordic) [reply src/index.ts]: I'm unsure but I would hypothesize it's only really needed for JITP association (if at all). It might be worth testing

**Review by gfleming-nordic — APPROVED** [review]

File: README.md [hunk 10]
```diff
@@ -11,27 +11,27 @@ This is an AWS IoT Thing simulator for nRF91. It shows how to use the Device API
 ### Most basic usage
 The most basic usage is just creating a device. For that you just need an API key. The device ID will be randomly generated and the rest of the necessary information (mqtt endpoint, cert, and mqtt prefix) will be pulled from the device API. 
 
-This will create a new device with AWS IoT, but it will not associate it to your account (use the `-a` flag) for that.
+The host defaults to the production environment (https://api.nrfcloud.com). **Nordic Semiconductor personnel**: if you want to use the simulator on our `dev`, `feature`, or `beta` environments, set a `API_HOST` env var to `https://api.[dev|feature|beta].nrfcloud.com` (you will need an AWS account for that environment). You can also set the `API_HOST` env var to the url of your sub account (ie `https://api.<user_id>.nrfcloud.com`) and the `stage` will be automatically set to `dev` (the stage is used for the job subscription topics for FOTA). You no longer need to set a `STAGE` variable, and it will be ignored.
```
Comment (lincolnthomas) [inline README.md]: Why is this here? This is a public repo, and a while ago we scrubbed out any reference to other stages, including here I thought. Remove it, we can describe usages for other hosts elsewhere in private docs like Confluence

**Review by lincolnthomas — CHANGES_REQUESTED** [review]

File: README.md (hunk 10 above)
Comment (sherretten) [reply README.md]: Ahh, this was added from a merge conflict. Apologies.

**Review by lincolnthomas — APPROVED** [review]
